text/entities/0256-entities-data-model.md

```diff
@@ -0,0 +1,765 @@
+
+When the entity's state is changed it is expected that the source will
+emit a new EntityState event with a fresh timestamp and full list of
+values of all other fields.
+
+Entity event producers are recommended to periodically emit events even
+if the Entity does not change. In this case the Type, Id and Attribute
+fields will remain the same, but a fresh Timestamp will be recorded in
+the event. Producing such events allows the system to be resilient to
+event losses. Even if some events are lost eventually the correct state
+of the Entity is more likely to be delivered to the final destination.
+Periodic sending of EntityState events also serves as a liveliness
+indicator (see below how it can be used in lieu of EntityDelete event).
+
+### EntityDelete Event
+
+EntityDelete event indicates that a particular entity is gone:
+
+<table>
+   <tr>
+    <td><strong>Field</strong>
+    </td>
+    <td><strong>Type</strong>
+    </td>
+    <td><strong>Description</strong>
+    </td>
+   </tr>
+   <tr>
+    <td>Timestamp
+    </td>
+    <td>nanoseconds
+    </td>
+    <td>The time when the entity was deleted. The time is measured by
+the origin clock. The field is required.
+    </td>
+   </tr>
+   <tr>
+    <td>Type
+    </td>
+    <td>
+    </td>
+    <td>See data model
+</td>
+   </tr>
+   <tr>
+    <td>Id
+    </td>
+    <td>
+    </td>
+    <td>See data model
+</td>
+   </tr>
+</table>
+
+Note that EntityDelete is optional and is not guaranteed to be sent when
+the entity is gone. Recipients of entity signals should be prepared to
+handle this situation by expiring entities that are no longer seeing
+EntityState events reported (i.e. treat the presence of EntityState
+events as a liveliness indicator).
+
+The expiration mechanism is based on the previously reported `Interval` field of
+EntityState event. The recipient can use this value to compute when to expect the next
+EntityState event and if the event does not arrive in a timely manner (plus some slack)
+it can consider the entity to be gone even if the EntityDelete event was not observed.
+
+## Entity Identification
+
+_This section is a supplementary guideline and is not part of logical data model._
+
+The data model defines the structure of the entity ID field. This section explains
+how the ID field is computed.
+
+### LID, GID and IDCONTEXT
+
+All entities have a local ID (LID) and a global ID (GID).
+
+The LID is unique in a particular identification context, but is not necessarily globally
+unique. For example a process entity's LID is its PID number and process start time.
+The (PID,StartTime) pair is unique only in the context of a host where the process runs
+(and the host in this case is the identification context).
+
+The GID of an entity is globally unique, in the sense that for the entire set of entities
+in a particular telemetry store no 2 entities exist that have the same GID value.
+
+The GID of an entity E is defined as:
+
+`GID(E) = { LID(E), GID(IDCONTEXT(E)) }`
```

Comment:
Something I’m not super clear on:

My understanding of the formula here suggests that it is recursive, in the sense that in order to calculate the GID for an entity, you will need to know the LID of the parent, and the parent’s parent, until you hit a GID.

Is it expected that the sender will always have enough available information/access to parent entity information to send a GID? What if it doesn't?

Reply:
>My understanding of the formula here suggests that it is recursive, in the sense that in order to calculate the GID for an entity, you will need to know the LID of the parent, and the parent’s parent, until you hit a GID.

That is exactly correct, it is recursive. I should probably also include the base case `GID(ROOT)={}` to make it clear.

>Is it expected that the sender will always have enough available information/access to parent entity information to send a GID? What if it doesn't?

No, this is not expected. It is explicitly called out on line 62 that this information may not be available in its entirety in one place. The sender may send an incomplete GID to the collection pipeline and elements of pipeline that have additional knowledge about entity's IDCONTEXT can add further elements to the ID. This is also known as "enrichment" and in the early version of the proposal we described how Collector does it [here](https://docs.google.com/document/d/1VUdBRInLEhO_0ABAoiLEssB1CQO_IcD5zDnaMEha42w/edit#heading=h.ij8yjheo645z).

Here is an example to visualize the process of computing a Process GID step by step (warning: no claims are made that Process will have exactly this ID attributes - the following diagram is just to illustrate the principle):
![image](https://github.com/open-telemetry/oteps/assets/4194920/e3f6b363-0c0b-4149-befd-a7afc8d0a416)

Reply:
Given that, since we know in advance that there will be cases where an entity is not uniquely identifiable from the data available to the sender, can this still be said to be true?

`Meaningful (especially human-readable) IDs are more valuable than random-generated IDs`

I would argue that a randomly generated UUID would be far more valuable than a meaningful id that does not uniquely distinguish the entity.

Reply:
@jsuereth - yes, to be clear I don't disagree with any of this **except in a context where an entity cannot be uniquely identified give available data**, in which case wouldn't we prefer to guarantee that it is uniquely identifiable?

Reply:
@HunterSherms you are absolutely right that exceptions exist. In fact we had to recently make this painful exception for [service.instance.id](https://github.com/open-telemetry/semantic-conventions/tree/main/docs/resource#service-experimental) field and recommend using a UUID for it.

The design principle of course is not absolute (none of the principles are) and is a merely a guide that helps make decisions. The principles are intentionally soft-worded and don't use the RFC-normative language (e.g. capitalized MUST or SHOULD) you see elsewhere.

Reply:
Excellent, thank you for the clarification!